### PR TITLE
Add text-conditioned in-painting functionality to Keras Stable Diffusion

### PR DESCRIPTION
**Pull request by bobqywei**
This PR adds in-painting support for the Keras port of Stable Diffusion following the RePaint method proposed in https://arxiv.org/abs/2201.09865. Also updated the URL and checksum of the model weights of the VAE Encoder.

Following are demo images:

Original Image with caption "a realistic photo of a chef cooking on mars.":
original

Mask Image:
image

Inpainted with caption "a realistic photo of a chef making a fruit salad.":
image

Inpainted with caption "a realistic photo of a chef with a boombox.":
image

### CODE REVIEW

**Comment by bobqywei** [comment]
@LukeWood here's the first draft

File: keras_cv/models/generative/stable_diffusion/image_encoder.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ def __init__(self, img_height=512, img_width=512, download_weights=True):
 
         if download_weights:
             image_encoder_weights_fpath = keras.utils.get_file(
-                origin="https://huggingface.co/fchollet/stable-diffusion/blob/main/vae_encoder.h5",
-                file_hash="f142c8c94c6853cd19d8bfb9c10aa762c057566f54456398beea6a70a639bf48",
+                origin="https://huggingface.co/fchollet/stable-diffusion/resolve/main/vae_encoder.h5",
+                file_hash="c60fb220a40d090e0f86a6ab4c312d113e115c87c40ff75d11ffcf380aab7ebb",
```
Comment (LukeWood) [inline keras_cv/models/generative/stable_diffusion/image_encoder.py:67]: Does the model not work with the original link + hash?

Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/image_encoder.py:67]: Perhaps we need to rebase?

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 2]
```diff
@@ -283,6 +283,144 @@ def generate_image(
...
+        num_resamples=1,
+        batch_size=1,
+        num_steps=25,
+        unconditional_guidance_scale=7.5,
+        diffusion_noise=None,
+        seed=None,
+    ):
+        """Generates an image based on encoded text.
+
+        The encoding passed to this method should be derived from
```
Comment (LukeWood) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:300]: it looks like the prompt is encoded internally in this method.

Lets update the docstring to match

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 3]
```diff
@@ -283,6 +283,144 @@ def generate_image(
...
+        image,
+        mask,
+        num_resamples=1,
+        batch_size=1,
+        num_steps=25,
+        unconditional_guidance_scale=7.5,
+        diffusion_noise=None,
+        seed=None,
+    ):
+        """Generates an image based on encoded text.
```
Comment (LukeWood) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:298]: `Inpaints a masked section of the provided image based on the provided prompt`

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 4]
```diff
@@ -283,6 +283,144 @@ def generate_image(
...
+            image: Tensor of shape (`batch_size`, `image_height`, `image_width`,
+            3) with RGB values in [0, 255]. When the batch is omitted, the same
+            image will be used as the starting image.
+            mask: Tensor of shape (`batch_size`, `image_height`, `image_width`)
+            with binary values 0 or 1. When the batch is omitted, the same mask
+            will be used on all images.
+            batch_size: number of images to generate. Default: 1.
+            num_steps: number of diffusion steps (controls image quality).
+                Default: 25.
+            unconditional_guidance_scale: float controling how closely the image
```
Comment (LukeWood) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:314]: "controlling"

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 5]
```diff
@@ -283,6 +283,144 @@ def generate_image(
...
+                should adhere to the prompt. Larger values result in more
+                closely adhering to the prompt, but will make the image noisier.
+                Default: 7.5.
+            diffusion_noise: Tensor of shape (`batch_size`, img_height // 8,
```
Comment (LukeWood) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:318]: start this description with (Optional) because the default is randomly generated, but we should still tell users tehy don't have to pass it.

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 6]
```diff
@@ -283,6 +283,144 @@ def generate_image(
...
+                img_width // 8, 4). Optional custom noise to seed the diffusion
+                process. When the batch axis is omitted, the same noise will be
+                used to seed diffusion for every generated image.
+            seed: integer which is used to seed the random generation of
+                diffusion noise, only to be specified if `diffusion_noise` is
+                None.
+        """
+        if diffusion_noise is not None and seed is not None:
+            raise ValueError(
+                "`diffusion_noise` and `seed` should not both be passed to "
```
Comment (LukeWood) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:329]: "Please pass either `diffusion_noise` or `seed` to `inpaint()`,`seed` is only used to generate diffusion "
                "noise when it's not already user-specified. "
f"Received `diffusion_noise={diffusion_noise}, seed={seed}"

**Review by LukeWood — CHANGES_REQUESTED** [review]
Overall this looks pretty good to me; I left a few questions for you.

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 7]
```diff
@@ -283,6 +283,144 @@ def generate_image(
...
+                    tf.expand_dims(diffusion_noise, axis=0), batch_size, axis=0
+                )
+            latent = diffusion_noise
+        else:
+            latent = self._get_initial_diffusion_noise(batch_size, seed)
+
+        # Iterative reverse diffusion stage
+        timesteps = tf.range(1, 1000, 1000 // num_steps)
+        alphas, alphas_prev = self._get_initial_alphas(timesteps)
+        progbar = keras.utils.Progbar(len(timesteps))
```
Comment (LukeWood) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:378]: Lets offer a way to turn this off to users;  my suggestions is an argument:`verbose`

**Comment by LukeWood** [comment]
I've also added @fchollet for suggestions as this method contains some somewhat complex logic.

File: keras_cv/models/generative/stable_diffusion/image_encoder.py (hunk 1 above)
Comment (bobqywei) [reply keras_cv/models/generative/stable_diffusion/image_encoder.py:67]: Rebase to master?

Comment (bobqywei) [reply keras_cv/models/generative/stable_diffusion/image_encoder.py:67]: Yeah this link + hash pair doesn't work for me for some reason?

Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/image_encoder.py:67]: Interesting, which one works?  c60fb220a40d090e0f86a6ab4c312d113e115c87c40ff75d11ffcf380aab7ebb or f142c8c94c6853cd19d8bfb9c10aa762c057566f54456398beea6a70a639bf48?

Lets just make sure we have the correct one in there.

**Review by LukeWood — APPROVED** [review]

**Comment by LukeWood** [comment]
@bobqywei if you run `./shell/format.sh` and add the updates then good to go from my end.

**Comment by LukeWood** [comment]
Thanks @bobqywei !  I'll run copybara soon so we can get to work!
